## Тестовое задание для маркетингового аналитика

# 1

In [1]:
import pandas as pd

### 1)

In [46]:
inst = pd.read_csv('Installs.csv')
v1 = pd.read_csv('cost_v1.csv')
v2 = pd.read_csv('cost_v2.csv')
both = pd.concat([v1,v2])
ms1 = both[both['MediaSource']=='MediaSource1']
ms2 = both[both['MediaSource']=='MediaSource2']

In [75]:
x = 0.0
length = len(ms1)
for i in range(length):
    i = ms1['Cost'].iloc[i]
    i = i.replace(',', '.')
    x += float(i)
z = x / len(inst.loc[inst['MediaSource']=='MediaSource1'])
print(f'CPI={z}')

CPI=0.7775701715271087


In [76]:
y = 0.0
length = len(ms2)
for i in range(length):
    i = ms2['Cost'].iloc[i]
    i = i.replace(',', '.')
    y += float(i)
z = y / len(inst.loc[inst['MediaSource']=='MediaSource2'])
print(f'CPI={z}')

CPI=9.084994071878475


In [77]:
paym = pd.read_csv('payments.csv')
inst = pd.read_csv('installs.csv')
paym = paym.merge(inst)
paym1 = paym.loc[paym['MediaSource']=='MediaSource1']
paym2 = paym.loc[paym['MediaSource']=='MediaSource2']
paym1

UserID PaymentRevenue   MediaSource  \
0     1673672646759-6010710452938637680    5,826125793  MediaSource1   
1     1673811886533-3308641358962027481    1,163524761  MediaSource1   
2     1673557631838-4075714709968288531    6,056467465  MediaSource1   
3     1673557631838-4075714709968288531    6,056467465  MediaSource1   
4     1673713438526-5393379852337099847    3,832763742  MediaSource1   
...                                 ...            ...           ...   
1003  1674984821480-3381443419258436996    12,02057553  MediaSource1   
1004  1675137689767-3192684835033167792    3,168388341  MediaSource1   
1005  1675137689767-3192684835033167792    3,168388341  MediaSource1   
1006  1675160441592-3459186411664614638    2,993603412  MediaSource1   
1007    1675106333473-55515122597745289    1,380200092  MediaSource1   

       CampaignID  Channel Country  Wifi  OSVersion  
0     19560308588  Network      BR  True         11  
1     19560308588   Search      BR  True         10  
2     19560308588  Network      NZ  True         10  
3     19560308588  Network      NZ  True         10  
4     19560308588   Search      MX  True         11  
...           ...      ...     ...   ...        ...  
1003  19560308588  Network      KR  True         11  
1004  19560308588  Network      PH  True          7  
1005  19560308588  Network      PH  True          7  
1006  19560305480  Network      JP  True         12  
1007  19560305480  Network      TR  True         10  

[894 rows x 8 columns]

In [80]:
length = len(paym1)
roas = 0
for i in range(length):
    i = paym['PaymentRevenue'].iloc[i]
    i = i.replace(',', '.')
    roas += float(i)
roas = (roas / x) * 100
print(f'ROAS={roas}')

ROAS=47.01945444748972


In [81]:
length = len(paym2)
roas = 0
for i in range(length):
    i = paym['PaymentRevenue'].iloc[i]
    i = i.replace(',', '.')
    roas += float(i)
roas = (roas / y) * 100
print(f'ROAS={roas}')

ROAS=24.67769562200672


### 2)

In [53]:
paym = pd.read_csv('payments.csv')
inst = pd.read_csv('installs.csv')

In [57]:
paym = paym.merge(inst)
paym = paym[paym['MediaSource']=='MediaSource2']
inst = inst[inst['MediaSource']=='MediaSource2']
paym

UserID PaymentRevenue   MediaSource  \
0    1673670959707-4461760122653175614           1,99  MediaSource2   
1    1673670959707-4461760122653175614           5,99  MediaSource2   
2    1673670959707-4461760122653175614           2,99  MediaSource2   
3    1673670959707-4461760122653175614           2,99  MediaSource2   
4    1673670959707-4461760122653175614           9,99  MediaSource2   
..                                 ...            ...           ...   
681    1673448265203-13427474715603838           1,99  MediaSource2   
682  1673436045529-6255423594941336299           9,99  MediaSource2   
683  1673190246471-5027966427991876824           4,99  MediaSource2   
684  1673409134243-2253051822699111766           1,99  MediaSource2   
685  1673173448844-7123897302277463823           4,99  MediaSource2   

           CampaignID     Channel Country   Wifi  OSVersion  
0    lQX9cNlleNjAtuzw    APPLOVIN      US   True         12  
1    lQX9cNlleNjAtuzw    APPLOVIN      US   True         12  
2    lQX9cNlleNjAtuzw    APPLOVIN      US   True         12  
3    lQX9cNlleNjAtuzw    APPLOVIN      US   True         12  
4    lQX9cNlleNjAtuzw    APPLOVIN      US   True         12  
..                ...         ...     ...    ...        ...  
681  lQX9cNlleNjAtuzw  IRONSOURCE      US  False         12  
682  lQX9cNlleNjAtuzw      VUNGLE      US  False         13  
683  lQX9cNlleNjAtuzw       UNITY      US   True         12  
684  lQX9cNlleNjAtuzw    APPLOVIN      US   True         13  
685  lQX9cNlleNjAtuzw  IRONSOURCE      US  False         13  

[686 rows x 8 columns]

In [85]:
con = (len(paym.loc[paym['Wifi']==True])/len(inst.loc[inst['Wifi']==True]))*100
print(f'Конверсия установивших через wifi={round (con, 2)}%')

Конверсия установивших через wifi=6.53%


In [87]:
con = len(paym.loc[paym['Wifi']==False])/len(inst.loc[inst['Wifi']==False])*100
print(f'Конверсия установивших не через wifi={round (con, 2)}%')

Конверсия установивших не через wifi=8.53%


*Достаточно ли данных, чтобы с 95% вероятностью утверждать, что влияние наличия
Wi-Fi имеет влияние на конверсию в платящего? Обоснуйте.*

Полученная информация указывает на то, что пользователи, устанавливающие приложение через wifi с большей вероятностью становятся платящими пользователями. Возможно, это связано с тем, что эта группа пользователей использует мобильный интернет, и, как правило, у них больше вероятность потратить деньги на приложение. Возможно, они более мобильны, или у них более высокий доход, что позволяет им использовать мобильный интернет для загрузки приложений и совершать внутриигровые покупки. Однако перед принятием окончательных выводов необходимо провести более глубокий анализ. Может быть, есть другие факторы, которые влияют на эти различия в конверсии. Кроме того, стоит учесть объем выборки данных, из которых были получены эти проценты. Если выборка была небольшой (а так и есть:) ), выводы могут быть не совсем точными.

### 3)

In [101]:
v1 = pd.read_csv('cost_v1.csv')
v2 = pd.read_csv('cost_v2.csv')
both = pd.concat([v1,v2])
both.loc[both['Country']=='US']

MediaSource        CampaignID Country        Cost Channel
87   MediaSource1       19560308588      US  318,121673     NaN
171  MediaSource1       19560305480      US  220,289754     NaN
434  MediaSource2  lQX9cNlleNjAtuzw      US  12260,1995     NaN

In [113]:
paym = pd.read_csv('payments.csv')
inst = pd.read_csv('installs.csv')
paym = paym.merge(inst)
paym.sort_values(by=['PaymentRevenue'], ascending=False).head(20)

UserID PaymentRevenue   MediaSource  \
23    1673670959707-4461760122653175614          99,99  MediaSource2   
1043   1672725758250-368411423781320192          99,99  MediaSource2   
1036   1672725758250-368411423781320192          99,99  MediaSource2   
1502  1674570331889-7473441113110627719          99,99  MediaSource2   
1204   1674003640748-700372152099542771           9,99  MediaSource2   
1347  1674460867228-7670623291692771055           9,99  MediaSource2   
1336  1674460867228-7670623291692771055           9,99  MediaSource2   
1335  1674460867228-7670623291692771055           9,99  MediaSource2   
1334  1674992672627-2449730529934400636           9,99  MediaSource2   
1326  1674992672627-2449730529934400636           9,99  MediaSource2   
1322  1672878546884-8504240120713449588           9,99  MediaSource2   
1320  1674347869067-4109476319833582658           9,99  MediaSource2   
1318  1674347869067-4109476319833582658           9,99  MediaSource2   
1306  1674347869067-4109476319833582658           9,99  MediaSource2   
1305  1674347869067-4109476319833582658           9,99  MediaSource2   
1293  1674392150767-6447681183379761331           9,99  MediaSource2   
1286  1674392150767-6447681183379761331           9,99  MediaSource2   
1282  1674148039448-9153813700456078433           9,99  MediaSource2   
1274  1674174840844-4011526515616317260           9,99  MediaSource2   
1264  1673758510405-2011308112947298959           9,99  MediaSource2   

            CampaignID     Channel Country   Wifi  OSVersion  
23    lQX9cNlleNjAtuzw    APPLOVIN      US   True         12  
1043  lQX9cNlleNjAtuzw    APPLOVIN      US   True         12  
1036  lQX9cNlleNjAtuzw    APPLOVIN      US   True         12  
1502  lQX9cNlleNjAtuzw  IRONSOURCE      US   True         12  
1204  lQX9cNlleNjAtuzw  IRONSOURCE      US   True          8  
1347  lQX9cNlleNjAtuzw    APPODEAL      US   True         12  
1336  lQX9cNlleNjAtuzw    APPODEAL      US   True         12  
1335  lQX9cNlleNjAtuzw    APPODEAL      US   True         12  
1334  lQX9cNlleNjAtuzw       FYBER      US  False         13  
1326  lQX9cNlleNjAtuzw       FYBER      US  False         13  
1322  lQX9cNlleNjAtuzw  IRONSOURCE      US  False         13  
1320  lQX9cNlleNjAtuzw      VUNGLE      US   True         11  
1318  lQX9cNlleNjAtuzw      VUNGLE      US   True         11  
1306  lQX9cNlleNjAtuzw      VUNGLE      US   True         11  
1305  lQX9cNlleNjAtuzw      VUNGLE      US   True         11  
1293  lQX9cNlleNjAtuzw       UNITY      US   True         13  
1286  lQX9cNlleNjAtuzw       UNITY      US   True         13  
1282  lQX9cNlleNjAtuzw  IRONSOURCE      US   True         12  
1274  lQX9cNlleNjAtuzw       UNITY      US  False         10  
1264  lQX9cNlleNjAtuzw      INMOBI      US   True          9

In [116]:
paym_sort = paym.sort_values(by=['PaymentRevenue'], ascending=False)
us = paym_sort.loc[paym_sort['Country'] == 'US']
ussum = 0
for i in range(len(us)):
    i = us['PaymentRevenue'].iloc[i]
    i = i.replace(',', '.')
    ussum += float(i)
ussum

7103.741788969886

In [123]:
paym.loc[paym['Country'] == 'US']

UserID PaymentRevenue   MediaSource  \
7     1673670959707-4461760122653175614           1,99  MediaSource2   
8     1673670959707-4461760122653175614           5,99  MediaSource2   
9     1673670959707-4461760122653175614           2,99  MediaSource2   
10    1673670959707-4461760122653175614           2,99  MediaSource2   
11    1673670959707-4461760122653175614           9,99  MediaSource2   
...                                 ...            ...           ...   
1575    1673448265203-13427474715603838           1,99  MediaSource2   
1576  1673436045529-6255423594941336299           9,99  MediaSource2   
1577  1673190246471-5027966427991876824           4,99  MediaSource2   
1578  1673409134243-2253051822699111766           1,99  MediaSource2   
1579  1673173448844-7123897302277463823           4,99  MediaSource2   

            CampaignID     Channel Country   Wifi  OSVersion  
7     lQX9cNlleNjAtuzw    APPLOVIN      US   True         12  
8     lQX9cNlleNjAtuzw    APPLOVIN      US   True         12  
9     lQX9cNlleNjAtuzw    APPLOVIN      US   True         12  
10    lQX9cNlleNjAtuzw    APPLOVIN      US   True         12  
11    lQX9cNlleNjAtuzw    APPLOVIN      US   True         12  
...                ...         ...     ...    ...        ...  
1575  lQX9cNlleNjAtuzw  IRONSOURCE      US  False         12  
1576  lQX9cNlleNjAtuzw      VUNGLE      US  False         13  
1577  lQX9cNlleNjAtuzw       UNITY      US   True         12  
1578  lQX9cNlleNjAtuzw    APPLOVIN      US   True         13  
1579  lQX9cNlleNjAtuzw  IRONSOURCE      US  False         13  

[722 rows x 8 columns]

In [125]:
inst.loc[inst['Country'] == 'US']

MediaSource        CampaignID     Channel Country   Wifi  \
132    MediaSource1       19560305480      Search      US   True   
290    MediaSource1       19560308588      Search      US   True   
323    MediaSource1       19560305480     Network      US   True   
359    MediaSource1       19560305480     Network      US   True   
420    MediaSource1       19560308588     Network      US   True   
...             ...               ...         ...     ...    ...   
22400  MediaSource2  lQX9cNlleNjAtuzw       FYBER      US   True   
22401  MediaSource2  lQX9cNlleNjAtuzw  IRONSOURCE      US  False   
22402  MediaSource2  lQX9cNlleNjAtuzw    APPLOVIN      US   True   
22403  MediaSource2  lQX9cNlleNjAtuzw  IRONSOURCE      US  False   
22404  MediaSource2  lQX9cNlleNjAtuzw       UNITY      US   True   

                                  UserID  OSVersion  
132    1674680568739-6822284685079598775          8  
290    1673779649413-1040532098514765506          6  
323    1674435705683-7937297099633863633          6  
359    1675025756649-5243684292360942905          6  
420    1673748691886-1069398316501351308          7  
...                                  ...        ...  
22400  1673919763777-9128193064855052089          8  
22401  1674114339859-3603234515028807846          8  
22402  1674446787782-9165755954283405579          8  
22403  1674836380753-8979378413452612436          8  
22404  1675170548683-3070412976466384047          8  

[3002 rows x 7 columns]

In [127]:
both.sort_values(by=['Cost'], ascending=False)

MediaSource   CampaignID Country       Cost Channel
204  MediaSource1  19560308588      JP  99,354037     NaN
150  MediaSource1  19560305480      AR  96,937839     NaN
339  MediaSource1  19560308588      MX  96,205985     NaN
11   MediaSource1  19560305480      CL  93,116949     NaN
44   MediaSource1  19560308588      IT  91,478102     NaN
..            ...          ...     ...        ...     ...
33   MediaSource1  19560308588      AQ          0     NaN
303  MediaSource1  19560305480      SJ          0     NaN
366  MediaSource1  19560308588      LS          0     NaN
309  MediaSource1  19560308588      SL          0     NaN
107  MediaSource1  19560308588      SH          0     NaN

[459 rows x 5 columns]

In [130]:
paym.loc[paym['Country'] == 'JP']

UserID PaymentRevenue   MediaSource  \
625   1674687564501-6402273028980797343    5,225483252  MediaSource1   
626   1674687564501-6402273028980797343    10,47285736  MediaSource1   
627   1674687564501-6402273028980797343    5,213324245  MediaSource1   
628   1674687564501-6402273028980797343    5,223375904  MediaSource1   
629   1674687564501-6402273028980797343     5,23378808  MediaSource1   
939   1675096662355-6094828701649063370    10,52346639  MediaSource1   
940   1675096662355-6094828701649063370    5,255846248  MediaSource1   
1006  1675160441592-3459186411664614638    2,993603412  MediaSource1   

       CampaignID  Channel Country  Wifi  OSVersion  
625   19560305480  Network      JP  True         13  
626   19560305480  Network      JP  True         13  
627   19560305480  Network      JP  True         13  
628   19560305480  Network      JP  True         13  
629   19560305480  Network      JP  True         13  
939   19560308588  Network      JP  True         12  
940   19560308588  Network      JP  True         12  
1006  19560305480  Network      JP  True         12

In [131]:
jp = paym_sort.loc[paym_sort['Country'] == 'JP']
jpsum = 0
for i in range(len(jp)):
    i = us['PaymentRevenue'].iloc[i]
    i = i.replace(',', '.')
    jpsum += float(i)
jpsum

439.92

Соединил таблицы установок и платящих, отсортировал по убыванию - US находится в топе платящих. Посмотрел на долю процента платящих людей из us - получилась половина, также 3002 среди всех установивших. Несмотря на то, что на кампании тратилось немало в сравнении с другими странами, ни одна страна не окупилась так сильно в процентном соотношении. Выводы сделать несложно - стоит продолжать вкладываться в us. А вообще, несколько советов относительно сделанных выводов датасета: провести исследования с целью определения своей целевой аудитории - это позволит вам создать более точные и персонализированные рекламные сообщения, которые будут привлекать больше интереса и вовлеченности потенциальных клиентов; изучить конкурентов в своей нише и анализировать их рекламные стратегии, использовать полученные знания для определения сильных и слабых сторон конкурентов и адаптирования своих рекламных приемов для достижения конкурентного преимущества; тестирование и оптимизация, то есть, провести A/B тестирование разных версий объявлений, заголовков, изображений и т.д. для определения наиболее эффективных вариантов.

# 2

1. Последний клик (Last Click):
   
   Преимущества: прост в реализаии и понимании, может быть эффективен для прямых маркетинговых каналов

   Критические ограничения: не учитывает другие маркетинговые каналы, которые могли также влиять на установку, не учитывает долгосрочные воздействия маркетинговых компаний
   

2. Линейная модель (Linear model):

   Преимущества: учитывает вклад всех маркетинговых каналов и действий, подходит для проектов с несколькими каналами привлечения

   Критические ограничения: может недооценить важность ключевых маркетинговых каналов, не учитывает временные различия во влиянии каналов
   

3. Атрибуция по времени (Time Decay):

   Преимущества: учитывает временную близость маркетинговых каналов к установке, позволяет учитывать долгосрочные и краткосрочные воздействия

   Критические ограничения: может недооценить ранние этапы привлечения, может быть солжно определить оптимальный временной интервал для учета
   

4. Модель атрибуции на основе данных (Data-Driven Attribution):

   Преимущества: использует данные о поведении пользователей для определения влияния маркетинговых каналов, может учитывать индивидуальные особенности проекта

   Критические ограничения: требует доступа к большому объему данных для точных результатов, может быть слоэно реализовать без специализированных инструментов и экспертизи в аналитике данных

# 3

Задача подразумевает случайный процесс с изменениями состояния и занимает область теории вероятностей и статистики. Поэтому точное решение может быть весьма сложным (постучимся к методу Монте-Карло). Прохождение уровня занимаетет 3м или 180с. Вероятность проигрыша при любой попытке составляет 75%, что приводит к тому, что на одну успешную попытку приходится в среднем 4 попытки (1/(1-0,75)). Таким образом, среднее время, затраченное на прохождение одного уровня, составляет 720с. Также после каждой попытки (будь то успешное прохождение уровня или поражение), игрок может взять паузу 10-30с с вероятностью 20% Предположим, что среднее время перерыва составляет 20 секунд. Тогда среднее время паузы после каждой игры составит 4с(20% от 20с). Таким образом, полное время, затрачиваемое на одну попытку (включая паузу), составляет 720 + 4 = 724с. В одном часе 3600 секунд, таким образом, предположительное количество уровней, которые игрок пройдет за один час, составит 3600 / 724 ≈ 4,97, это скорее всего будет округлено до 5 уровней. Это предположительные и усредненные значения. Реальное количество уровней может значительно варьироваться в зависимости от специфичных обстоятельств каждой игры.

# 4

1. "Among Us": Эта кооперативная игра для мобильных устройств разработана InnerSloth. Игроки должны работать вместе, чтобы поддерживать корабль в исправности, отправляясь на разные миссии, в то время как «импостеры» пытаются помешать им и «убить» их.

2. "Call of Duty: Mobile": Это игра в жанре шутер от первого лица, разработанная Activision для мобильных устройств. Игроки могут сражаться в различных режимах, включая мультиплеер и королевскую битву, используя различное оружие и навыки.

3. "Genshin Impact": Это игра в стиле аниме, разработанная miHoYo. Игроки исследуют открытый мир, борются с различными врагами, решают головоломки и выполняют квесты, а также прокачивают и улучшают персонажей.

4. "Mobile Legends: Bang Bang": Это мобильная игра в жанре MOBA, где две команды из пяти игроков борются друг с другом, стараясь уничтожить базу противника. Игроки могут выбирать различных героев с уникальными способностями.

5. "Pokеmon Go": Это игра в жанре AR (дополненной реальности), разработанная Niantic. Игроки двигаются по реальному миру, пытаясь отловить виртуальных покемонов на своих смартфонах и участвуя в боях.